File: lib/repository/user/my_sql.py
from datetime import datetime;
from interfaces.user_repository import IUserRepository;

import os;
import uuid;
import mysql.connector;

# Implements IPostRepository interface for connecting to a MySQL database.
# See interfaces/post_repository for method documentation


class UserMySQLRepository(IUserRepository):
    _table_name = "users";

    """
    @param (object) self
    @returns (None)
    """

    def __init__(self):
        self._db_connection = mysql.connector.connect(
          host = os.getenv("DATABASE_HOSTNAME"),
          user = os.getenv("DATABASE_USERNAME"),
          password = os.getenv("DATABASE_PASSWORD"),
          database = os.getenv("DATABASE_NAME")
        );


    def create(self, doc):
        db_cursor = self._db_connection.cursor();
        my_uuid = str(uuid.uuid4());
        created_date = datetime.now();
        query = ("INSERT INTO users (id, handle, email_address, motto, phone_number, first_name, last_name, created_date) VALUES (%s, %s, %s, %s, %s, %s, %s, %s)");

        doc.update({"id": my_uuid, "created_date": created_date});
        db_cursor.execute(query, (
            my_uuid,
            doc["handle"],
            doc["email_address"],
            doc["motto"],
            doc["phone_number"],
            doc["first_name"],
            doc["last_name"],
            created_date
        ));
        self._db_connection.commit();
        db_cursor.close();

        return { "id": my_uuid, "created_date": str(created_date) };


    def find_one(self, id):
        db_cursor = self._db_connection.cursor();
        query = ("SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users WHERE id = '{}'".format(id));

        db_cursor.execute(query);
        result = db_cursor.fetchall();

        return list(map(lambda p: self.on_read_user(p), result));


    def find_one_by_email(self, email_address):
        db_cursor = self._db_connection.cursor();
        query = ("SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users WHERE email_address = '{}'".format(email_address));

        db_cursor.execute(query);
        result = db_cursor.fetchall();

        return list(map(lambda p: self.on_read_user(p), result));


    def find_one_by_handle(self, handle):
        db_cursor = self._db_connection.cursor();
        query = ("SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users WHERE handle = '{}'".format(handle));

        db_cursor.execute(query);
        result = db_cursor.fetchall();

        return list(map(lambda p: self.on_read_user(p), result));


    def find_all(self):
        user_list = []
        db_cursor = self._db_connection.cursor();
        db_cursor.execute("SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users");
        result = db_cursor.fetchall();

        for user in result:
            user_list.append(self.on_read_user(user));

        return user_list;


    def edit_motto(self, id, text):
        db_cursor = self._db_connection.cursor();
        last_modified = datetime.now();
        query = ("UPDATE users SET motto = '{}', last_modified = '{}' WHERE id = '{}'".format(text, last_modified, id));

        db_cursor.execute(query);
        self._db_connection.commit();
        db_cursor.close();

        return { "id": id, "last_modified": str(last_modified) };


    def edit_name(self, id, doc):
        db_cursor = self._db_connection.cursor();
        last_modified = datetime.now();
        query = ("UPDATE users SET first_name = '{}', last_name = '{}', last_modified = '{}' WHERE id = '{}'".format(
            doc["first_name"],
            doc["last_name"],
            last_modified,
            id
        ));

        db_cursor.execute(query);
        self._db_connection.commit();
        db_cursor.close();

        return { "id": id, "last_modified": str(last_modified) };


    def edit_phone_number(self, id, phone_number):
        db_cursor = self._db_connection.cursor();
        last_modified = datetime.now();
        query = ("UPDATE users SET phone_number = '{}', last_modified = '{}' WHERE id = '{}'".format(
            phone_number,
            last_modified,
            id
        ));

        db_cursor.execute(query);
        self._db_connection.commit();
        db_cursor.close();

        return { "id": id, "last_modified": str(last_modified) };


    def create_subscription(self, current_user_id, target_user_id):
        try:
            db_cursor = self._db_connection.cursor();
            insert_subscription = ("INSERT INTO user_followers (user_id, follower_id) VALUES(%s, %s)");
            increment_follower_count = ("UPDATE users SET follower_count = follower_count + 1 WHERE id = '{}'".format(target_user_id));

            db_cursor.execute(insert_subscription, (target_user_id, current_user_id));
            db_cursor.execute(increment_follower_count);
            self._db_connection.commit();

        except mysql.connector.Error as transaction_error:
            self._db_connection.rollback();
            print("TransactionError: {}".format(transaction_error));
        finally:
            db_cursor.close();


    def subscription_exists(self, current_user_id, target_user_id):
        db_cursor = self._db_connection.cursor();
        query = ("SELECT COUNT(*) FROM user_followers WHERE user_id = '{}' AND follower_id = '{}'".format(target_user_id, current_user_id));
        db_cursor.execute(query);

        result = db_cursor.fetchall();
        db_cursor.close();

        return result[0][0];


    def get_subscribers_of(self, current_user_id):
        user_list = [];
        db_cursor = self._db_connection.cursor();
        query = ("SELECT users.id, users.handle, users.email_address, users.motto, users.is_verified, users.first_name, users.last_name, users.follower_count, users.created_date, user_followers.* FROM user_followers JOIN users ON user_followers.follower_id = users.id WHERE user_followers.user_id = '{}'".format(current_user_id));
        db_cursor.execute(query);

        result = db_cursor.fetchall();
        db_cursor.close();

        for user in result:
            user_list.append(self.on_read_user(user));

        return user_list;


    def get_user_subscriptions(self, current_user_id):
        user_list = [];
        db_cursor = self._db_connection.cursor();
        query = ("SELECT users.*, user_followers.* FROM user_followers JOIN users ON user_followers.user_id = users.id WHERE user_followers.follower_id = '{}'".format(current_user_id));
        db_cursor.execute(query);

        result = db_cursor.fetchall();
        db_cursor.close();

        for user in result:
            user_list.append(self.on_read_user(user));

        return user_list;


    def delete(self, id):
        pass;


    def on_read_user(self, record):
        return {
          "id": record[0],
          "handle": record[1],
          "email_address": record[2],
          "motto": record[3],
          "is_verified": record[4],
          "first_name": record[5],
          "last_name": record[6],
          "follower_count": record[7],
          "created_date": record[8]
        }

####PostMySQLRepository###

File: src/lib/repository/user/mysql.js
/*Implements IUserRepository interface for connecting to a MySQL database.
See interfaces/user-repository for method documentation*/

const uuid = require("uuid");
const { promisify } = require("util");

/**
 * @implements {IUserRepostory}
 * @param {Object} databaseConnector - object with methods for connecting to a database 
 */

function UserMySQLRepository(databaseConnector) {

    this.create = async function(doc) {
        const createdDate = new Date().toISOString();
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const id = uuid.v4();

        const sql = `INSERT INTO users (id, handle, email_address, motto, phone_number, first_name, last_name, created_date) VALUES ("${id}", "${doc.handle}", "${doc.emailAddress}", "${doc.motto}", "${doc.phoneNumber}", "${doc.firstName}", "${doc.lastName}", "${createdDate}");`;

        const result = await runQuery(sql);
        connection.release();

        return { "id": id, "createdDate": createdDate };
    }


    this.createUserPassword = async function(userEmailAddress, password) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));

        const sql = `INSERT INTO user_credentials (user_email_address, user_password) VALUES ("${userEmailAddress}", "${password}");`;

        await runQuery(sql);
        connection.release();
    }


    this.getUserPassword = async function(userEmailAddress) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const sql = `SELECT user_password FROM user_credentials WHERE user_email_address = "${userEmailAddress}";`;

        const [result] = await runQuery(sql);
        connection.release();

        return result.user_password;
    }


    this.findOneById = async function(id) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const sql = `SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users WHERE id = '${id}'`;
        const result = await runQuery(sql);
        return result.map((u) => onReadUser(u));
    }


    this.findOneByEmail = async function(emailAddress) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));

        const sql = `SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users WHERE email_address = '${emailAddress}'`;

        const result = await runQuery(sql);
        connection.release();

        return result.map((u) => onReadUser(u));
    }


    this.findOneByHandle = async function(handle) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const sql = `SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users WHERE handle = '${handle}'`;
        const result = await runQuery(sql);
        connection.release();

        return result.map((u) => onReadUser(u));
    }


    this.findAll = async function() {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const sql = `SELECT id, handle, email_address, motto, is_verified, first_name, last_name, follower_count, created_date FROM users`;
        const result = await runQuery(sql);
        connection.release();

        return result.map((u) => onReadUser(u));
    }


    this.editMotto = async function(id, text) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const lastModified = new Date().toISOString();
        const sql = `UPDATE users SET motto = '${text}', last_modified = '${lastModified}' WHERE id = '${id}'`;

        const result = await runQuery(sql);
        connection.release();

        return { "id": id, "lastModified": lastModified };
    }


    this.editName = async function(id, doc) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const lastModified = new Date().toISOString();
        const sql = `UPDATE users SET first_name = '${doc.firstName}', last_name = '${doc.lastName}', last_modified = '${lastModified}' WHERE id = '${id}'`;

        const result = await runQuery(sql)
        connection.release();

        return { "id": id, "lastModified": lastModified };

    }


    this.editPhoneNumber = async function(id, phoneNumber) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const lastModified = new Date().toISOString();
        const sql = `UPDATE users SET phone_number = '${phoneNumber}', last_modified = '${lastModified}' WHERE id = '${id}'`;

        const result = await runQuery(sql);
        connection.release();

        return { "id": id, "lastModified": lastModified };
    }


    this.createSubscription = async function(currentUserId, targetUserId) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const insertSubscriptionSql = `INSERT INTO user_followers (user_id, follower_id) VALUES('${targetUserId}', '${currentUserId}')`;
        const incrementFollowerCountSql = `UPDATE users SET follower_count = follower_count + 1 WHERE id = '${targetUserId}'`;

        connection.beginTransaction(async(err) => {
            try {
                await runQuery(insertSubscriptionSql);
                await runQuery(incrementFollowerCountSql);
                connection.commit();
            }
            catch (e) {
                console.error(e);
                connection.rollback();
            }
        });
        connection.release();
    }


    this.removeSubscription = async function(currentUserId, targetUserId) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const removeSubscription = `DELETE FROM user_followers WHERE follower_id = '${currentUserId}'`;
        const decrementFollowerCount = `UPDATE users SET follower_count = follower_count - 1 WHERE id = '${targetUserId}'`;

        connection.beginTransaction(async(err) => {
            try {
                await runQuery(removeSubscription);
                await runQuery(decrementFollowerCount);
                connection.commit();
            }
            catch (e) {
                console.error(e);
                connection.rollback();
            }
        });
        connection.release();
    }


    this.subscriptionExists = async function(currentUserId, targetUserId) {
        /*
        TODO: Figure out what this is supposed to do. Seems redudant since there's already a user.isFollowing and user.follows method
        db_cursor = self._db_connection.cursor();
        query = ("SELECT COUNT(*) FROM user_followers WHERE user_id = '{}' AND follower_id = '{}'".format(target_user_id, current_user_id));
        db_cursor.execute(query);

        result = db_cursor.fetchall();
        db_cursor.close();

        return result[0][0];
        */
    }


    this.getSubscribersOf = async function(currentUserId) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const sql = `SELECT users.id, users.handle, users.email_address, users.motto, users.is_verified, users.first_name, users.last_name, users.follower_count, users.created_date, user_followers.* FROM user_followers JOIN users ON user_followers.follower_id = users.id WHERE user_followers.user_id = '${currentUserId}'`;
        const result = await runQuery(sql);

        return result.map((u) => onReadUser(u));
    }


    this.getUserSubscriptions = async function(currentUserId) {
        const connection = await databaseConnector.getConnection();
        const runQuery = promisify(connection.query.bind(connection));
        const sql = `SELECT users.*, user_followers.* FROM user_followers JOIN users ON user_followers.user_id = users.id WHERE user_followers.follower_id = '${currentUserId}'`;

        const result = await runQuery(sql);
        return result.map((u) => onReadUser(u));
    }

    this.deleteUser = function(id) {
        return
    }


    function onReadUser(record) {
        return {
            id: record.id,
            handle: record.handle,
            emailAddress: record.email_address,
            motto: record.motto,
            isVerified: record.is_verified,
            firstName: record.first_name,
            lastName: record.last_name,
            followerCount: record.follower_count,
            createdDate: record.created_date
        }
    }
}

/*UserMySQLRepository*/

module.exports = UserMySQLRepository;
